### ISSUE TITLE
Required children are not enforced with TSX

### ISSUE THREAD

**Issue by agubler** [microsoft/TypeScript#38108]
I couldn't find a specific issue related to this behaviour, apologies if one already exists. When defining TSX namespace typings for children adding an empty set of braces, `{}`, seems to fulfil the typings even when children are required.

**TypeScript Version:**  3.8.3

**Search Terms:**

tsx, children, required

**Code**

```tsx
declare global {
    namespace JSX {
        type Element = any;
        interface ElementAttributesProperty {
            __properties__: {};
        }
        interface IntrinsicElements {
            [key: string]: string;
        }
        interface ElementChildrenAttribute {
            __children__: {};
        }
    }
}

// Mock React
export function React(): any {}

interface MockComponentInterface {
	new (): {
        __properties__: { bar?: number } & { __children__: () => number };
	};
}

declare const MockComponent: MockComponentInterface;

// correctly infers that children are required and reports error
<MockComponent></MockComponent>;

// expecting this to error as the child is not passed, only an empty `{}`
<MockComponent>{}</MockComponent>;

// correctly infers the types when a child is passed
<MockComponent>{() => ''}</MockComponent>;
<MockComponent>{() => 1}</MockComponent>;
```

**Expected behavior:**

Expected the `MockComponent` to report an error when an empty set of braces are passed as children.

**Actual behavior:**

No error occurs

**Playground Link:** https://www.typescriptlang.org/play/?jsx=2&ssl=17&ssc=23&pln=17&pc=68#code/CYUwxgNghgTiAEBzCB7ARlC8DeAoeB8AdlALYgDOADlGAgFIDKAGjvoRwC4CeVCAohBDkineAF54UItwDc7DgQCWokDABmtAUJGcAgp04wlaAK6dKABRgo+MHm0VP4AfRdUbdzkspuAXDgAvvLO8IEKiioWGlrwAJKixkQUSmCCwiCiFI6hhADaANYg3AEURiqIALql5USIIc7hoVFqmnTw6boAwgAWShDAcEQG5WYWObmuLmB9A0P+QQ1OTYTh4bgA9BvwALIoYAXwAEogtJy4IAAeVCj28OqmRGDeKETHp88AFACUAdLcQVwuBaMXaewOXRQpBuREynAS0TaCDwAEhYQB3eA-AJ4UJuDy2NTeXwuHHwDAwAD8ASIplIaDUYXgADIcFMZv1BpkFj8JAA+Yh0hkwMLyFHBXDrUCQWAIMCvMq7fYFSHQ15wgLglVQmFwhGtLTyTbbeUwODPCAAlTqNTZTg9KBiDlzTJSODwOAAR1MSjgwCkRH9cBu9myahsMFwAB4tardaI+VGNrGdeqE0atvArnxnhV4PalHaUFmzbcpHaenLZv7C8QUGIaBQKCBgAAaeCvS0BrPQhwAA2wgT70ZTathCcHSdH8c4fIzJtu5s4XettvzlfzvEo8HRlbeUHgzpr2UbzeAI+VcbTs+wvPEAoA5A-AlPL6nx7P5DG32O4Xzb98-LwAAjC+yY-jOc5AA

**Related Issues:** N/A

**Comment by canonic-epicure** [comment]
Just stumbled upon this one too. It seems the typechecking of the `children` property has multiple problems. Found 2 issues so far, none of the error messages make sense. [Playground link](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictPropertyInitialization=false&declaration=false&ssl=29&ssc=23&pln=29&pc=27#code/PQKhAIAECsGcA9wFkCeApAGgZQHQGMAnAUwEMAXIgUQBsiBbIgOzPBGACh2ATIvak4uADm1APYAjEtXABvduAXhGJBrAAOJPEXBosGWfMVGyKNdpr0mLI4oC84DHWoWGzQzfAmz4F1YDC-LCwdg5Ovm7uNgCWzEQEAGaa5rSuZACCZGQEUeIArhSwAAoEomYEJgYeHmolarCRRgC+nFXgMRQJST4p-gAWUdRcxIwZWTn52nKtRnj9gwByojz1rc1N7M3sRPBqouVKKkTqXaiYWJWK27v78bmMeGRRoozghKQU4SwAFA2KZCRCeaHcAALnAsDGjCEABpwOQxnkCqDZOAANrgADWRBQyIh2Sh4AAurjIUJwI1YTgqbMBkMmMivniYmSAD6hZw9ZgASlRhMiXORjg5lmYFw8wGAbToak5-0ezzawUYohYsCiQkYUXiUTwJDcrWIZFyBBejCIAHd2Z8vlzIs1Nuw+CQglbZWLwDVSsFpsiyP0VooaQslkcPGDGaTwGyhZ8ecSPPZeRtOE6XVhmbQ-HMuJ9wNsKIwuMEY26poHs4tlmG0eDI-Gfa1k47AsEACrm0RZ2nDXP5phF10ilhlhRBriV0M2MHoktD2Gz1JEhvTJsS8Ct-06PSeAEAcmCu7HE9gu49tTzO14ZGCXm0u9RTKhhNP5v6s3AxAAjrkosRgnRcmoR4ZW0MdhlhRFwGeagcRIcEM1A7NwHNZ0z1EAA3KIeC4HAvgAJgAdgAFgAVltPBnghPNqAARnAewAB50yhTNs0+AA+GRwAAImdLhuPJBjgGYkQiC7QYOM4NdilKOIKl3T06lPKJ-xUtUCRiTxTDvLijxDWBpwXKx5zCWVCQAbnJU9IK-H9iC4NoXlvcBd3bTtsx7WVd1wwiiJo8jKJYIhqDw+jwAYtzxLpRhJKMBiuCidD2KEhKksieLEuS4BUvY9ghMijymA4oA)

```ts
/** @jsx MyJXS.createElement */

declare global {
    namespace JSX {
        type Element        = XmlElement
        type ElementClass   = XmlElement

        interface ElementAttributesProperty {
            props
        }

        interface ElementChildrenAttribute {
            childNodes
        }
    }
}

export namespace MyJXS {
    export function createElement (
        tagName : string, attributes : { [ key : string ] : string }, ...children : (string | XmlElement)[]
    ) : XmlElement {
        // implementation is not significant
        return new XmlElement()
    }
}


class XmlElement {
    props           : this
    childNodes      : (string | XmlElement)[]       = []
}

class SingleChildElement extends XmlElement {
    childNodes      : [ string ]                    
}

class TwoChildrenElement extends XmlElement {
    childNodes      : [ XmlElement, XmlElement ]                    
}

// This JSX tag's 'childNodes' prop expects type '[string]' which requires multiple children, but only a single child was provided.(2745)
const el1 = <SingleChildElement>{ "asd" }</SingleChildElement>

// Property 'props' is missing in type '{ childNodes: [XmlElement, XmlElement]; }' but required in type 'TwoChildrenElement'.(2741)
const el2 = <TwoChildrenElement>
    <div></div>
    <div></div>
</TwoChildrenElement>
```

**Comment by Lonli-Lokli** [comment]
Not sure if it same case, but children are not well typed in React at all

```
interface CaseProp<T> {
    when: T,
    children: React.ReactNode
}

export function Switch<T>({
    parent,
    children
}: {parent: T, children: ReactElement<CaseProp<T>> | ReactElement<CaseProp<T>>[]}) {
    return (
        <>
            <div>Parent is {parent}</div>
            {children}
        </>
    );
}

export function Case<T>({when, children}: CaseProp<T>) {
    return (
        <>
            <div>Value is {when}</div>
            {children}
        </>
    )
}

export function App() {
    const unionValue: 't1' | 't2' = 't1';
    return (
        <Switch parent={unionValue}>
            <Case when="unknown" >
                <div>Case child</div>
            </Case>
        </Switch>
    )
}
```

[playground](https://www.typescriptlang.org/play?&q=365#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIgDoB6AOwgBMkB9ESgVwBskBnYgARgE8xXipUGfAG4AsAChQkWHADeiQTACizEElLwAvnGy4CJclVoNmbTjz4D0MERInANSKJnRI4AYRQskABRxgAHgAVAD45CThIuAB3AAt1AC44IIAaCKi0WOBGCgFSJORrQkKMADlKJAlNO3EkAA9pLHpSDGAIUjgAZWjgGEzgkIAKWXTIsBQ8mDTxKLhM7Nz1KqTZccmk1LmsnLyCxRUkNQ0Az28-CEDQsIAfBWsDo5gTr19-AZCAbQBdTQBKcJmUQEMHoUA6g1GszgARCkKhUQCFGAADcQj4Jup4MAWHI1pjNAFiEjUXD4XJ5jt1NVAfDCbCaXBfmJxNSJPVGjpmq12h4XgNhnF1Cktgs8pokqdXhcBv8RgzgaDwaTIjDlVDESiQgA1FCMehubFyQWkAlEzVq2ayCmLE1qumQ35VGrs6BNFowNodACCYDAg1lkLQ7RY8Ganp1eqQSQA5DAAIzRuC3WMAJkTAF44LGE8zZgqwXAIQzZgFur1MnA8Rp07Iw+0I-rNPSyerJTF4qR0-hmgBrcjRUj4ODNltkjWotvWwnEkej6HESWz9XEMt9WJLx0soA)
